packages/cardano-services-client/src/HandleProvider/KoraLabsHandleProvider.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  Cardano,
+  HandleProvider,
+  HandleResolution,
+  HealthCheckResponse,
+  NetworkInfoProvider,
+  ProviderError,
+  ProviderFailure,
+  ResolveHandlesArgs
+} from '@cardano-sdk/core';
+
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { IHandle } from '@koralabs/handles-public-api-interfaces';
+import axios, { AxiosAdapter, AxiosInstance } from 'axios';
+
+export interface KoraLabsHandleProviderDeps {
+  serverUrl: string;
+  networkInfoProvider: NetworkInfoProvider;
+  adapter?: AxiosAdapter;
+  policyId: Cardano.PolicyId;
+}
+
+export const toHandleResolution = ({
+  apiResponse,
+  tip,
+  policyId
+}: {
+  apiResponse: IHandle;
+  tip: Cardano.Tip;
+  policyId: Cardano.PolicyId;
+}): HandleResolution => ({
+  handle: apiResponse.name,
+  hasDatum: apiResponse.hasDatum,
+  policyId,
+  resolvedAddresses: {
+    cardano: Cardano.PaymentAddress(apiResponse.resolved_addresses.ada)
+  },
+  resolvedAt: {
+    hash: tip.hash,
+    slot: tip.slot
+  }
+});
+
+export class KoraLabsHandleProvider implements HandleProvider {
+  private axiosClient: AxiosInstance;
+  private networkInfoProvider: NetworkInfoProvider;
+  policyId: Cardano.PolicyId;
+
+  constructor({ serverUrl, networkInfoProvider, adapter, policyId }: KoraLabsHandleProviderDeps) {
+    this.networkInfoProvider = networkInfoProvider;
+    this.axiosClient = axios.create({
+      adapter,
+      baseURL: serverUrl
+    });
+    this.policyId = policyId;
+  }
+
+  async resolveHandles(args: ResolveHandlesArgs): Promise<HandleResolution[]> {
+    try {
+      const tip = await this.networkInfoProvider.ledgerTip();
+      const response = await Promise.all(
+        args.handles.map((handle) => this.axiosClient.get<IHandle>(`/handles/${handle}`))
+      );
+      return response.map(({ data: apiResponse }) => toHandleResolution({ apiResponse, policyId: this.policyId, tip }));
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        if (error.request) {
+          throw new ProviderError(ProviderFailure.ConnectionFailure, error, error.code);
+        }
+        throw new ProviderError(ProviderFailure.Unhealthy, error, `Failed to resolve handles due to: ${error.message}`);
+      }
+      throw new ProviderError(ProviderFailure.Unknown, error, 'Failed to resolve handles');
```

Comment:
In the `try` block there is at least one call which can throw a `ProviderError`, a check as follows is required before the catch all error in order to avoid it produces an `Unknown` `ProviderError` with another `ProviderError` as its `innerError`.

```suggestion
      if (error instanceof ProviderError) throw error;
      throw new ProviderError(ProviderFailure.Unknown, error, 'Failed to resolve handles');
```

Reply:
dde35b14797987 should address this